src/cheri_insts.sail

```diff
@@ -32,6 +32,108 @@
 /*  SUCH DAMAGE.                                                          */
 /*========================================================================*/
 
+/* Capability versions of mode-dependent instructions */
+
+union clause ast = AUIPCC : (bits(20), regidx)
+/*!
+ * Capability register *cd* is replaced with the contents of **PCC**, with the
+ * **address** replaced with **PCC**.**address** $+$ *imm* $\times$ 4096.
+ */
+function clause execute AUIPCC(imm, cd) = {
+  let off : xlenbits = EXTS(imm @ 0x000);
+  let (representable, newCap) = setCapAddr(PCC, PC + off);
+  if representable then
+    C(cd) = newCap
+  else
+    C(cd) = invalidateCap(newCap);
+  RETIRE_SUCCESS
+}
+
+union clause ast = CJAL : (bits(21), regidx)
+/*!
+ * Capability register *cd* is replaced with the next instruction's **PCC** and
+ * sealed as a sentry. **PCC**.**address** is incremented by *imm*.
+ *
+ * ## Exceptions
+ *
+ * An exception is raised if:
+ *   - **PCC**.**address** $+$ *imm* $\lt$ **PCC**.**base**.
+ *   - **PCC**.**address** $+$ *imm* $+$ min_instruction_bytes $\gt$ **PCC**.**top**.
+ *   - **PCC**.**address** $+$ *imm* is unaligned, ignoring bit 0.
+ */
+function clause execute(CJAL(imm, cd)) = {
+  let off : xlenbits = EXTS(imm);
+  let newPC = PCC.address + off;
+  if not(inCapBounds(PCC, newPC, min_instruction_bytes())) then {
+    handle_cheri_cap_exception(CapEx_LengthViolation, PCC_IDX);
+    RETIRE_FAIL
+  } else if newPC[1] == bitone & ~(haveRVC()) then {
+    handle_mem_exception(newPC,  E_Fetch_Addr_Align());
+    RETIRE_FAIL
+  } else {
+    let (success, linkCap) = setCapAddr(PCC, nextPC); /* Note that nextPC accounts for compressed instructions */
+    assert(success, "Link cap should always be representable.");
+    assert(not (isCapSealed(linkCap)), "Link cap should always be unsealed");
+    C(cd) = sealCap(linkCap, to_bits(cap_otype_width, otype_sentry));
+    nextPC = newPC;
+    RETIRE_SUCCESS
+  }
+}
+
+union clause ast = CJALR : (bits(12), regidx, regidx)
+/*!
+ * Capability register *cd* is replaced with the next instruction's **PCC** and
+ * sealed as a sentry. **PCC** is replaced with the value of capability
+ * register *cs1* with its **address** incremented by *imm* and the 0th bit of
+ * its **address** set to 0, and is unsealed if it is a sentry.
+ *
+ * ## Exceptions
+ *
+ * An exception is raised if:
+ *   - *cs1*.**tag** is not set.
+ *   - *cs1* is sealed and is not a sentry.
+ *   - *cs1* is a sentry and *imm* $\ne$ 0.
+ *   - *cs1*.**perms** does not grant **Permit_Execute**.
+ *   - *cs1*.**address** $+$ *imm* $\lt$ *cs1*.**base**.
+ *   - *cs1*.**address** $+$ *imm* $+$ min_instruction_bytes $\gt$ *cs1*.**top**.
+ *   - *cs1*.**base** is unaligned.
+ *   - *cs1*.**address** $+$ *imm* is unaligned, ignoring bit 0.
+ */
+function clause execute(CJALR(imm, cs1, cd)) = {
```

Comment:
I would have expected the arguments to be `imm, cd, cs1` or `cd, cs1, imm` but it matches the base instructions so I guess that's consistent.

Reply:
Yeah... sail-riscv puts them in encoding order, we put them generally in assembly order. For AUIPCC we followed AUIPC, so I did the same for CJAL[R]. We might want to reconsider that choice for them though.